### PR TITLE
[Perf] lower memory usage when loading multi-label datasets.

### PR DESCRIPTION
**Pull request by Miles-Ma**
Hi. I found mmclassification consumed too much RAM. The image below is RAM usages of self-implemented code (top) and mmclassification (bottom) when training on 1M, 500 classes multi-label dataset. This is deadly when training larger datasets like [Open Image](https://storage.googleapis.com/openimages/web/factsfigures.html).

comp1

In this PR, I modified `MultiLabelDataset` and its child class `VOC`, and now it consumes less RAM when training on multi-label datasets. The image below shows performance improvement.

comp2

### CODE REVIEW

**Comment by Miles-Ma** [comment]
The inefficient RAM usage is mainly caused by multiple worker processes of `DataLoader`. According to document of [`torch.multiprocessing`](https://pytorch.org/docs/stable/multiprocessing.html), tensors in `Dataset` will be copied to shared memory and shared with all dataloader workers, but numpy arrays are not. So, changing numpy arrays (i.e. labels) to tensors will save quite a lot RAM. But that's not enough, because copying massive tensors will reach limitation of max file handles, so I stack these tensors into a large one and keep an index of label for each sample.

**Comment by mzr1996** [comment]
Thx for your contribution!
We will review it recently

**Comment by Miles-Ma** [comment]
Updated multi-label evaluation test in tests/test_dataset.py, because now `data_infos` is `dict` instead of `list`.

File: mmcls/datasets/voc.py [hunk 1]
```diff
@@ -53,16 +55,29 @@ def load_annotations(self):
                 else:
                     labels.append(label)
 
-            gt_label = np.zeros(len(self.CLASSES))
+            gt_label = torch.zeros(len(self.CLASSES), dtype=torch.int8)
```
Comment (mzr1996) [inline mmcls/datasets/voc.py:58]: Is it necessary to use tensor instead of array?

File: mmcls/datasets/multi_label.py [hunk 2]
```diff
@@ -9,19 +11,38 @@
...
+        gt_label_index = self.data_infos['samples'][idx]['gt_label_index']
+        gt_label = self.data_infos['all_gt_labels'][gt_label_index]
+        cat_ids = torch.where(gt_label == 1)[0]
         return cat_ids
 
+    def prepare_data(self, idx):
+        results = copy.deepcopy(self.data_infos['samples'][idx])
+        gt_label_index = results.pop('gt_label_index')
+        results['gt_label'] = self.data_infos['all_gt_labels'][gt_label_index]
+        return self.pipeline(results)
```
Comment (mzr1996) [inline mmcls/datasets/multi_label.py:41]: Modify `prepare_data` will influence all classes inherited from `MultiLabelDataset`, which will cause serious BC-breaking.
Consider creating a new class or use an option to disable this behavior by default?

**Review by mzr1996 — CHANGES_REQUESTED** [review]

**Comment by mzr1996** [comment]
Please merge the latest master branch and handle the conflict.

File: mmcls/datasets/voc.py [hunk 3]
```diff
@@ -53,16 +55,29 @@ def load_annotations(self):
...
 
-            info = dict(
+            sample = dict(
                 img_prefix=self.data_prefix,
                 img_info=dict(filename=filename),
-                gt_label=gt_label.astype(np.int8))
-            data_infos.append(info)
+                gt_label_index=sample_cnt)
+            samples.append(sample)
+            sample_cnt += 1
```
Comment (mzr1996) [inline mmcls/datasets/voc.py:70]: Consider using `enumerate(img_ids)`

**Review by mzr1996 — CHANGES_REQUESTED** [review]

File: mmcls/datasets/voc.py (hunk 1 above)
Comment (Miles-Ma) [reply mmcls/datasets/voc.py]: Yes, it is the key. Only tensors are shared among dataloader workers by `torch.multiprocessing`. Numpy arrays, however, are deep copied to each workers, which consumes a lot of RAM.

File: mmcls/datasets/multi_label.py (hunk 2 above)
Comment (Miles-Ma) [reply mmcls/datasets/multi_label.py]: I now revert all changes on multi_label.py, and do this optimization only on voc.py. That will not affect datasets. Other developers may take voc.py as an example for optimization.

File: mmcls/datasets/voc.py [hunk 4]
```diff
@@ -53,16 +55,29 @@ def load_annotations(self):
...
 
+        # Stack tensors to accelerate copying to shared memory,
+        # also avoid being limited by max number of file descriptors.
+        # If there is a way to know total sample number from the beginning,
+        # we'd better pre-allocate memory of the whole tensor instead of
+        # stacking a list of tensors.
+        all_gt_labels = torch.stack(all_gt_labels)
```
[the remark this answers is not in the document]
Comment (Miles-Ma) [reply mmcls/datasets/voc.py]: You are right. Code updated.

File: mmcls/datasets/voc.py [hunk 5]
```diff
@@ -53,16 +88,19 @@ def load_annotations(self):
...
-            info = dict(
+            sample = dict(
                 img_prefix=self.data_prefix,
                 img_info=dict(filename=filename),
-                gt_label=gt_label.astype(np.int8))
-            data_infos.append(info)
+                gt_label_index=index)
+            samples.append(sample)
 
+        data_infos = dict(
```
Comment (mzr1996) [inline mmcls/datasets/voc.py:102]: `data_infos` should be `List[Dict]` instead of `Dict`, it's a protocol, please don't modify it.

**Review by mzr1996 — CHANGES_REQUESTED** [review]
Hello, here are some changes recommended, please keep in contact if you have any questions about them.

File: mmcls/datasets/voc.py [hunk 6]
```diff
@@ -53,16 +88,19 @@ def load_annotations(self):
...
+            all_gt_labels[index, labels_difficult] = -1
+            all_gt_labels[index, labels] = 1
 
-            info = dict(
+            sample = dict(
                 img_prefix=self.data_prefix,
                 img_info=dict(filename=filename),
-                gt_label=gt_label.astype(np.int8))
-            data_infos.append(info)
+                gt_label_index=index)
```
[the remark this answers is not in the document]
Comment (Miles-Ma) [reply mmcls/datasets/voc.py:99]: Yes, this is not beneficial to reducing mem cost, but I still need this to deal with "insufficient-file-descriptors" problem.

That is, because of the default sharing strategy `file_descriptor` (see section "Sharing strategies" in the [doc](https://pytorch.org/docs/stable/multiprocessing.html) or the following figure), sharing large number of tensors will likely exhaust file handles. That means if you have `N` training samples, then you should always set your `ulimit -n` bigger than `N`, that's not realistic for large datasets. Thus I stack `gt_label`s to a single tensor `all_gt_label` to avoid that. The problem exists and this solution works, I have verified them. Another possible solution is to change sharing strategy to `file_system` but I'm not sure whether it brings other side effects so I didn't change that.

capture

File: mmcls/datasets/voc.py (hunk 5 above)
Comment (Miles-Ma) [reply mmcls/datasets/voc.py:102]: I understand. Can I instead make `all_gt_labels` a field of `VOC` instance then? I need a place to store common things that can be accessed by all samples in `prepare_data` method.

File: mmcls/datasets/voc.py [hunk 7]
```diff
@@ -24,15 +25,49 @@ def __init__(self, **kwargs):
...
         """Load annotations.
 
         Returns:
             list[dict]: Annotation info from XML file.
         """
-        data_infos = []
+        samples = []
         img_ids = mmcv.list_from_file(self.ann_file)
-        for img_id in img_ids:
+        all_gt_labels = torch.zeros(len(img_ids), len(self.CLASSES),
```
[the remark this answers is not in the document]
Comment (Miles-Ma) [reply mmcls/datasets/voc.py:68]: Yes, that's why I opened this PR : )

[the remark this answers is not in the document]
Comment (mzr1996) [reply mmcls/datasets/voc.py:68]: Sorry, I missed [this comment](https://github.com/open-mmlab/mmclassification/pull/315#issuecomment-863730888).

**Comment by mzr1996** [comment]
@jshilong Any update on this PR?
